### PR TITLE
WIP: refactoring for GGUF support

### LINKED ISSUE #60: Migrate to GGUF
This will also resolve #32 and #49

### PR DESCRIPTION
**Pull request by monatis**
closes #60, #49, #32

This is still WIP and not yet for anything yet.
I hope to finish it by Sunday.

- [x] Rewrite the conversion script for GGUF support.
- [x] Output text-only, vision-only and two-tower models according to the CLI argument.
- [x] Allow overriding mean / std values for image processing.
- [x] Load GGUF file and infer with it (WIP).
- [x] Re-implement the quantization function.
- [x] Support f32, q5_0, q5_1 and q8_0 models in addition to f16, q4_0 and q4_1.
- [x] Handle memory buffer allocation logic for new text-only and vision-only models.
- [x] Test and upload different CLIP models in new GGUF format.
- [x] Update Python bindings.

### CODE REVIEW

**Comment by monatis** [comment]
I set this ready for review although it will require some tests with different models and fixes in memory allocation for text-only and vision-only model variants. 

I'll do tests and make some code polishing over the coming days. Any test and feedback will be much appreciated btw.

**Comment by monatis** [comment]
I'm preparing the PR for merging later today. Added a shell script for bulk model conversion. Another script will do quantization in bulk. Then uploading models + updating readme and merging.

Python binding will require an update in its automatic model selection as now the smallest model in a repo is a text-only model, but I'll leave it to another PR.

**Comment by monatis** [comment]
Also updated the Python bindings to work with the new GGUF format and find pre-converted models uploaded to HF.

I think this is good to merge now.

File: README.md [hunk 1]
```diff
@@ -19,19 +19,22 @@ This repo is aimed at powering useful applications based on such models on compu
...
+
 ## Note about image preprocessing
 
 PIL uses a two-pass convolutions-based bicubic interpolation in resizing with antialiasing applied. In Pytorch, antialiasing is optional. It needs some extra attention to implement this preprocessing logic that matches their results numerically. However, I found that linear interpolation is also good enough for both comparison of different embeddings from this implementation and also comparison of an embedding from this implementation and another one from Transformers. So let's use it until we craft a proper bicubic interpolation.
 
 ## Preconverted Models
 
-Preconverted Models can be found in [HuggingFace Repositories tagged with `clip.cpp`](https://huggingface.co/models?other=clip.cpp).
+Preconverted Models can be found in [HuggingFace Repositories tagged with `clip.cpp-gguf`](https://huggingface.co/models?other=clip.cpp).
```
Comment (Green-Sky) [inline README.md]: you sure this is a good idea? i can take down /reupload my repos too.

also, you only changed the text, not the link

**Comment by Green-Sky** [comment]
file names are somewhat lenghty now, with "ggml-model" "ggml-text-model" and "ggml-vision-model". feels like the "ggml" here is redundant (.gguf).

File: README.md (hunk 1 above)
Comment (monatis) [reply README.md]: I don't have a strong opinion in this, but I just wanted to close this PR to jump to other tasks (implement BERT and large multimodal models). I only uploaded a few models --my internet speed is not so good. If you'd like to re-upload models I'd only be happy :D and new shell scripts might help you with conversions and quantizations.

**Comment by monatis** [comment]
yes, it might be {full|text|vision}-model-{ftype}.gguf

**Pull request merged**